include/openrave/kinbody.h

```diff
@@ -3703,6 +3705,10 @@ class OPENRAVE_API KinBody : public InterfaceBase
 
     std::vector<GrabbedPtr> _vGrabbedBodies; ///< vector of grabbed bodies
 
+    /// Quick lookup map of body name -> indexed into _vGrabbedBodies of the matching grab entry
+    using MapGrabbedBodyNameIndex = std::unordered_map<std::string, std::vector<GrabbedPtr>::size_type>;
+    MapGrabbedBodyNameIndex _mapGrabbedBodyNameIndex;
```

Comment:
@rschlaikjer Maintaining any more members as cache than there is is already scary. Anyway we can do this by better ordering _vGrabbedBodies? Perhaps the vector can be ordered?
Also, I'm wondering if we really need `_vGrabbedBodies` to be a vector of shared pointers. If we design an efficient swap function, perhaps it'll be pretty quick to insert and/or remove elements from the sorted vector (which doesn't happen as much as querying anyway..)

Reply:
IIRC initially I was just going to just turn `_vGrabbedBodies` into a map to avoid having a layer of indirection, but there exist APIs that seem to depend on the ordering of the members of `_vGrabbedBodies` (e.g. `GetGrabbedBody(int index)`), so it seemed like mutating the order could cause problems for existing code. 

Is the concern with cache the memory overhead? Or keeping it consistent?

I'm not sure how much the shared pointer to `Grabbed` is a performance issue; those are internal and generally swapped / referenced. The performance issue was locking the weak pointer to a kinbody _inside_ grabbed to fetch the name.

Reply:
Not sure that form of the function is so important. If it means we can optimize, I would remove it.

Also, perhaps worthwhile to consider using environmentBodyIndex instead of pointers? Any grabbed body is guaranteed to be in the environment, so its index shouldn't change, and can query its pointer from the environment pretty easily.

Reply:
Here's a quick version with the removal of GetGrabbedBody(int):
https://github.com/rdiankov/openrave/compare/rs-optimize-reset-grabbed-novector

It requires additional PRs (rb #2292, cc #1042) to account for the API change.

Reply:
Okay, what was the purpose of `uintptr_t _grabbedBodyPtrId;`? I didn't see it used anywhere, and I'm also not a big fan of using pointers as Ids...

Also, why not make the key the environmentBodyIndex? Why do we need name if we can quickly use an integer to get to the Grabbed object?
And `GetGrabbed(std::vector<KinBodyPtr>& vbodies)` can also be returning a vector of environmentBodyIndex's


I think the more we move away from shared pointers, dynamically allocated std::strings, and sets/maps, the better it is for the entire systems.
If we're going to re-write, might as well do it with that in mind: how to reduce the dynamic memory allocation/object creation as much as possible.

Reply:
> _grabbedBodyPtrId

oops, that's a remnant of a route I didn't go down. Dead code.

> Also, why not make the key the environmentBodyIndex? 

The inital reason was that when regrabbing we need to do a lot of looking up by name, since that's what the GrabbedInfo structs contain - if the map is based on index, then we need a double-lookup of name -> body -> env index.

What are the rules for environmentBodyIndex invalidation? Is it ever possible for a grabbed body's index to change without the grab relationship changing?

Reply:
Why is the lookup by name necessary?  And `environmentBodyIndex` should not be changing for actively grabbed bodies.

Reply:
We have to do a lookup by name with the GrabbedInfos only contain a name, and we want to match them to existing Grabbed structs to avoid destroying/recreating all grabs on reset grabbed. 

Modified the example map version to use env body index and merged into this branch.